mountpoint-s3/src/prefetch/parquet_prefetch.rs

```diff
@@ -0,0 +1,181 @@
+use crate::sync::Arc;
+use bytes::Bytes;
+use futures::pin_mut;
+use futures::StreamExt;
+use mountpoint_s3_client::types::ETag;
+use mountpoint_s3_client::ObjectClient;
+use parquet::errors::ParquetError;
+use parquet::file::footer::decode_footer;
+use parquet::file::metadata::ParquetMetaData;
+use parquet::file::reader::{ChunkReader, Length};
+use std::collections::HashMap;
+use std::io::{Read, Seek, SeekFrom};
+use tracing::debug_span;
+use tracing::trace;
+
+use super::PrefetchReadError;
+
+pub async fn read_parquet_metadata<Client>(
+    client: Arc<Client>,
+    bucket: &str,
+    key: &str,
+    if_match: &ETag,
+    total_size: u64,
+) -> Result<Bytes, PrefetchReadError<Client::ClientError>>
+where
+    Client: ObjectClient + Send + Sync + 'static,
+{
+    let chunk_reader = ParquetS3ChunkReader {
+        client,
+        bucket: bucket.to_owned(),
+        key: key.to_owned(),
+        etag: if_match.clone(),
+        size: total_size,
+    };
+
+    parse_metadata_raw(&chunk_reader).map_err(|_| PrefetchReadError::GetRequestTerminatedUnexpectedly)
```

Comment:
Hm, so we're blocking the thread here until the data is loaded from s3 and parsed, that is too expensive. We need to do the s3 request in asynchronous context so other futures may make progress while we wait for a response from s3.

I'd want us to be `await`ing s3 request(s) here instead of jumping to synchronous functions and blocking there.

Do we know in advance what part of s3 object we need to download to access the footer? May be by making two s3 requests, say "read footer offset at a predefine place in file` and next "read footer"?

Reply:
Resolved this in the new implementation ^.